File: src/fields/neural.cpp
#include <mitsuba/core/properties.h>
#include <mitsuba/render/field.h>

#include <sstream>
#include <vector>

NAMESPACE_BEGIN(mitsuba)

namespace {

FieldValueType parse_field_value_type(std::string_view value) {
    if (value == "Float")
        return FieldValueType::Float;
    if (value == "Spectrum")
        return FieldValueType::Spectrum;
    if (value == "Color3")
        return FieldValueType::Color3;
    if (value == "Array2")
        return FieldValueType::Array2;
    if (value == "Array3")
        return FieldValueType::Array3;
    if (value == "Features")
        return FieldValueType::Features;

    Throw("Invalid field output type \"%s\".", value);
}

FieldDomain parse_field_domain(std::string_view value) {
    if (value == "Surface")
        return FieldDomain::Surface;
    if (value == "Interaction")
        return FieldDomain::Interaction;
    if (value == "SurfaceAndInteraction")
        return FieldDomain::SurfaceAndInteraction;

    Throw("Invalid field domain \"%s\".", value);
}

} // namespace

/**!

.. _field-neuralfield:

Neural field (:monosp:`neuralfield`)
------------------------------------

.. pluginparameters::

 * - domain
   - |string|
   - Interaction records accepted by the field: ``Surface``,
     ``Interaction``, or ``SurfaceAndInteraction``. (Default: ``Surface``)

 * - out_type
   - |string|
   - Semantic type of the predicted output: ``Float``, ``Spectrum``,
     ``Color3``, ``Array2``, ``Array3``, or ``Features``. (Default:
     ``Color3``)

 * - out_dim
   - |integer|
   - Number of output channels. This is required for ``Features`` outputs and
     should match the fixed semantic dimension for other output types.

 * - args_dim
   - |integer|
   - Number of optional argument channels accepted by the field. A value of 0
     means that direct callers should not pass extra arguments. (Default: 0)

 * - encoding
   - |field|
   - Nested encoding field that maps interaction records to feature channels.
     Examples include :monosp:`hashgridencoding`,
     :monosp:`permutoencoding`, and :monosp:`sinusoidalencoding`.

 * - decoder
   - |string|
   - Decoder type applied to the encoded features: ``neural`` or ``linear``.
     (Default: ``neural``)

 * - hidden_size
   - |integer|
   - Hidden layer width used by the neural decoder.

 * - num_layers
   - |integer|
   - Number of decoder layers.

This plugin evaluates a learned field by combining an optional nested encoding
with a decoder network. It is intended for direct learned-field users and for
plugins such as :monosp:`neuralbsdf` that pass richer context through optional
argument channels.

Neural and cooperative-vector-backed implementations may be JIT-only. They
must reject unsupported scalar variants during construction with a clear error
message naming the plugin and supported variants.

*/

template <typename Float, typename Spectrum>
class NeuralField final : public Field<Float, Spectrum> {
public:
    MI_IMPORT_BASE(Field)
    MI_IMPORT_TYPES()

    using FloatStorage = typename Base::FloatStorage;
    using Args         = typename Base::Args;
    using Array2f      = typename Base::Array2f;
    using Array3f      = typename Base::Array3f;
    using Array6f      = typename Base::Array6f;

    NeuralField(const Properties &props) : Base(props) {
        m_domain = parse_field_domain(props.get<std::string_view>("domain", "Surface"));
        m_out_type = parse_field_value_type(props.get<std::string_view>("out_type", "Color3"));
        m_out_dim = props.get<uint32_t>("out_dim", 0);
        m_args_dim = props.get<uint32_t>("args_dim", 0);
        m_decoder = std::string(props.get<std::string_view>("decoder", "neural"));
        props.mark_queried("hidden_size");
        props.mark_queried("num_layers");

        if (props.has_property("encoding"))
            m_encoding = props.get<ref<Base>>("encoding");
    }

    FieldValueType out_type() const override { return m_out_type; }
    FieldDomain domain() const override { return m_domain; }
    uint32_t out_dim() const override { return m_out_dim; }
    uint32_t args_dim() const override { return m_args_dim; }

    bool supports_scalar() const override { return false; }
    bool supports_jit() const override { return true; }
    bool supports_surface_queries() const override {
        return m_domain == FieldDomain::Surface ||
               m_domain == FieldDomain::SurfaceAndInteraction;
    }
    bool supports_interaction_queries() const override {
        return m_domain == FieldDomain::Interaction ||
               m_domain == FieldDomain::SurfaceAndInteraction;
    }

    FloatStorage eval(const SurfaceInteraction3f &, Args, Mask) const override {
        NotImplementedError("eval");
    }

    FloatStorage eval(const Interaction3f &, Args, Mask) const override {
        NotImplementedError("eval");
    }

    Float eval_1(const SurfaceInteraction3f &, Args, Mask) const override {
        NotImplementedError("eval_1");
    }

    Float eval_1(const Interaction3f &, Args, Mask) const override {
        NotImplementedError("eval_1");
    }

    Color3f eval_color3(const SurfaceInteraction3f &, Args, Mask) const override {
        NotImplementedError("eval_color3");
    }

    Color3f eval_color3(const Interaction3f &, Args, Mask) const override {
        NotImplementedError("eval_color3");
    }

    Array2f eval_array2(const SurfaceInteraction3f &, Args, Mask) const override {
        NotImplementedError("eval_array2");
    }

    Array2f eval_array2(const Interaction3f &, Args, Mask) const override {
        NotImplementedError("eval_array2");
    }

    Array3f eval_array3(const SurfaceInteraction3f &, Args, Mask) const override {
        NotImplementedError("eval_array3");
    }

    Array3f eval_array3(const Interaction3f &, Args, Mask) const override {
        NotImplementedError("eval_array3");
    }

    UnpolarizedSpectrum eval_spec(const SurfaceInteraction3f &, Args, Mask) const override {
        NotImplementedError("eval_spec");
    }

    UnpolarizedSpectrum eval_spec(const Interaction3f &, Args, Mask) const override {
        NotImplementedError("eval_spec");
    }

    Array6f eval_array6(const SurfaceInteraction3f &, Args, Mask) const override {
        NotImplementedError("eval_array6");
    }

    Array6f eval_array6(const Interaction3f &, Args, Mask) const override {
        NotImplementedError("eval_array6");
    }

    void eval_n(const SurfaceInteraction3f &, Float *, uint32_t, Args, Mask) const override {
        NotImplementedError("eval_n");
    }

    void eval_n(const Interaction3f &, Float *, uint32_t, Args, Mask) const override {
        NotImplementedError("eval_n");
    }

    void traverse(TraversalCallback *cb) override {
        if (m_encoding)
            cb->put("encoding", m_encoding, ParamFlags::Differentiable);
    }

    void parameters_changed(const std::vector<std::string> &) override { }

    std::string to_string() const override {
        std::ostringstream oss;
        oss << "NeuralField[" << std::endl
            << "  out_dim = " << m_out_dim << "," << std::endl
            << "  args_dim = " << m_args_dim << "," << std::endl
            << "  decoder = \"" << m_decoder << "\"" << std::endl
            << "]";
        return oss.str();
    }

    MI_DECLARE_CLASS(NeuralField)

private:
    FieldDomain m_domain = FieldDomain::Surface;
    FieldValueType m_out_type = FieldValueType::Color3;
    uint32_t m_out_dim = 0;
    uint32_t m_args_dim = 0;
    std::string m_decoder;
    ref<Base> m_encoding;

    MI_TRAVERSE_CB(Base, m_encoding)
};

MI_EXPORT_PLUGIN(NeuralField)
NAMESPACE_END(mitsuba)

File: src/python/python/fields/neural.py
from __future__ import annotations # Delayed parsing of type annotations

import drjit as dr
import drjit.nn as nn

from .common import (
    _field_domain,
    _field_value_type,
    _field_value_type_name,
    _args_as_channels,
    _network_dtype,
    _require_jit,
    _semantic_dim,
    _surface_record,
    _validate_args,
)


def _make_neural_field(mi):
    class NeuralField(mi.Field):
        def __init__(self, props):
            super().__init__(props)
            _require_jit(mi, "neuralfield")

            self._domain = _field_domain(mi, props.get("domain", "Surface"))
            self._out_type = _field_value_type(mi, props.get("out_type", "Color3"))
            self._out_dim = int(props.get("out_dim", 0))
            self._args_dim = int(props.get("args_dim", 0))
            self._decoder = str(props.get("decoder", "neural")).lower()
            self._hidden_size = int(props.get("hidden_size", 16))
            self._num_layers = int(props.get("num_layers", 2))

            if self._decoder not in ("neural", "linear"):
                raise RuntimeError(
                    'neuralfield: decoder must be "neural" or "linear".'
                )
            if self._hidden_size <= 0:
                raise RuntimeError("neuralfield: hidden_size must be positive.")
            if self._num_layers < 0:
                raise RuntimeError("neuralfield: num_layers must be non-negative.")
            if self._args_dim < 0:
                raise RuntimeError("neuralfield: args_dim must be non-negative.")

            expected_dim = _semantic_dim(mi, self._out_type)
            if self._out_type == mi.FieldValueType.Features:
                if self._out_dim <= 0:
                    raise RuntimeError(
                        "neuralfield: out_dim for Features outputs must be positive."
                    )
            elif self._out_dim == 0:
                self._out_dim = expected_dim
            elif self._out_dim != expected_dim:
                raise RuntimeError(
                    f"neuralfield: out_dim ({self._out_dim}) does not match "
                    f"metadata out_type={_field_value_type_name(mi, self._out_type)} "
                    f"(expected {expected_dim})."
                )

            self._encoding = props.get("encoding", None)
            if self._encoding is not None:
                if (
                    self._encoding.out_type() != mi.FieldValueType.Features
                    or self._encoding.out_dim() <= 0
                ):
                    raise RuntimeError(
                        "neuralfield: encoding must be a Features field with "
                        "positive out_dim."
                    )
                if self._encoding.args_dim() != 0:
                    raise RuntimeError(
                        "neuralfield: encoding fields may not require args."
                    )
                if (
                    self.supports_surface_queries()
                    and not self._encoding.supports_surface_queries()
                ):
                    raise RuntimeError(
                        "neuralfield: encoding field does not support Surface "
                        "queries required by the selected domain."
                    )
                if (
                    self.supports_interaction_queries()
                    and not self._encoding.supports_interaction_queries()
                ):
                    raise RuntimeError(
                        "neuralfield: encoding field does not support "
                        "Interaction queries required by the selected domain."
                    )

            tensor_dtype, scalar_dtype = _network_dtype(
                mi, props.get("precision", "fp16"), plugin_name="neuralfield"
            )
            self._feature_dim = (
                5
                + self._args_dim
                + (self._encoding.out_dim() if self._encoding is not None else 0)
            )

            layers = [nn.Cast(scalar_dtype)]
            if self._decoder == "linear":
                layers.append(nn.Linear(self._feature_dim, self._out_dim))
            else:
                in_features = self._feature_dim
                for _ in range(self._num_layers):
                    layers.append(nn.Linear(in_features, self._hidden_size))
                    layers.append(nn.LeakyReLU())
                    in_features = self._hidden_size
                layers.append(nn.Linear(in_features, self._out_dim))

            network = nn.Sequential(*layers).alloc(
                tensor_dtype,
                size=self._feature_dim,
                rng=dr.rng(seed=int(props.get("seed", 0))),
            )
            self.network_weights, self._network = nn.pack(
                network, layout="training"
            )
            self._network_weight_count = dr.width(self.network_weights)

        def out_type(self):
            return self._out_type

        def domain(self):
            return self._domain

        def out_dim(self):
            return self._out_dim

        def args_dim(self):
            return self._args_dim

        def supports_scalar(self):
            return False

        def supports_jit(self):
            return True

        def supports_surface_queries(self):
            return self._domain in (
                mi.FieldDomain.Surface,
                mi.FieldDomain.SurfaceAndInteraction,
            )

        def supports_interaction_queries(self):
            return self._domain in (
                mi.FieldDomain.Interaction,
                mi.FieldDomain.SurfaceAndInteraction,
            )

        def _validate_surface_query(self):
            if not self.supports_surface_queries():
                raise RuntimeError(
                    "neuralfield: domain mismatch, field does not support "
                    "Surface queries."
                )

        def _validate_interaction_query(self):
            if not self.supports_interaction_queries():
                raise RuntimeError(
                    "neuralfield: domain mismatch, field does not support "
                    "Interaction queries."
                )

        def _features(self, record, args, active):
            is_surface = _surface_record(record)
            if is_surface:
                self._validate_surface_query()
                uv_x, uv_y = record.uv.x, record.uv.y
            else:
                self._validate_interaction_query()
                uv_x, uv_y = mi.Float(0), mi.Float(0)

            _validate_args("neuralfield", self._args_dim, args)
            arg_channels = _args_as_channels(args)
            features = [
                record.p.x,
                record.p.y,
                record.p.z,
                uv_x,
                uv_y,
            ]
            features.extend(arg_channels)

            if self._encoding is not None:
                encoded = self._encoding.eval_n(
                    record, self._encoding.out_dim(), active=active
                )
                features.append(encoded)

            return features

        def _eval_coop(self, record, args, active):
            return self._network(nn.CoopVec(*self._features(record, args, active)))

        def _validate_output(self, expected_type, expected_dim: int, method: str):
            if self._out_type != expected_type or self._out_dim != expected_dim:
                raise RuntimeError(
                    f"neuralfield::{method}(): incompatible output metadata "
                    f"(out_type={_field_value_type_name(mi, self._out_type)}, "
                    f"out_dim={self._out_dim})."
                )

        def eval(self, record, args=None, active=True):
            if args is None and self._args_dim == 0:
                if self._out_type == mi.FieldValueType.Float and self._out_dim == 1:
                    value = self.eval_1(record, args=[], active=active)
                    return mi.UnpolarizedSpectrum(value)
                if self._out_type == mi.FieldValueType.Spectrum:
                    return self.eval_spec(record, args=[], active=active)
                if (
                    self._out_type == mi.FieldValueType.Color3
                    and self._out_dim == 3
                ):
                    value = self.eval_color3(record, args=[], active=active)
                    if mi.is_monochromatic:
                        return mi.UnpolarizedSpectrum(mi.luminance(value))
                    if mi.is_spectral:
                        raise RuntimeError(
                            "neuralfield::eval(): cannot convert Color3 field "
                            "output to a spectral value."
                        )
                    return value
                if (
                    self._out_type == mi.FieldValueType.Array3
                    and self._out_dim == 3
                ):
                    value = self.eval_array3(record, args=[], active=active)
                    color = mi.Color3f(value.x, value.y, value.z)
                    if mi.is_monochromatic:
                        return mi.UnpolarizedSpectrum(mi.luminance(color))
                    if mi.is_spectral:
                        raise RuntimeError(
                            "neuralfield::eval(): cannot convert Array3 field "
                            "output to a spectral value."
                        )
                    return color
                raise RuntimeError(
                    "neuralfield::eval(): output is not spectrum-compatible; "
                    "use eval_n()."
                )
            return mi.ArrayXf(self._eval_coop(record, args, active)) & active

        def eval_1(self, record, args=None, active=True):
            result = list(self._eval_coop(record, args, active))

            if self._out_type == mi.FieldValueType.Float and self._out_dim == 1:
                value = mi.Float(result[0])
            elif self._out_type == mi.FieldValueType.Spectrum:
                value = dr.mean(mi.UnpolarizedSpectrum(result))
            elif (
                self._out_type in (mi.FieldValueType.Color3, mi.FieldValueType.Array3)
                and self._out_dim == 3
            ):
                value = mi.luminance(mi.Color3f(result[0], result[1], result[2]))
            else:
                raise RuntimeError(
                    "neuralfield::eval_1(): incompatible output metadata "
                    f"(out_type={_field_value_type_name(mi, self._out_type)}, "
                    f"out_dim={self._out_dim})."
                )

            return dr.select(active, value, 0)

        def eval_color3(self, record, args=None, active=True):
            self._validate_output(mi.FieldValueType.Color3, 3, "eval_color3")
            return mi.Color3f(self._eval_coop(record, args, active)) & active

        def eval_array2(self, record, args=None, active=True):
            self._validate_output(mi.FieldValueType.Array2, 2, "eval_array2")
            result = list(self._eval_coop(record, args, active))
            return mi.Point2f(result[0], result[1]) & active

        def eval_array3(self, record, args=None, active=True):
            self._validate_output(mi.FieldValueType.Array3, 3, "eval_array3")
            result = list(self._eval_coop(record, args, active))
            return mi.Vector3f(result[0], result[1], result[2]) & active

        def eval_spec(self, record, args=None, active=True):
            expected = dr.size_v(mi.UnpolarizedSpectrum)
            self._validate_output(mi.FieldValueType.Spectrum, expected, "eval_spec")
            return (
                mi.UnpolarizedSpectrum(self._eval_coop(record, args, active))
                & active
            )

        def eval_array6(self, record, args=None, active=True):
            self._validate_output(mi.FieldValueType.Features, 6, "eval_array6")
            result = list(self._eval_coop(record, args, active))
            return mi.ArrayXf(
                [dr.select(active, mi.Float(result[i]), 0) for i in range(6)]
            )

        def eval_n(self, record, count, args=None, active=True):
            if int(count) != self._out_dim:
                raise RuntimeError(
                    f"neuralfield::eval_n(): count ({int(count)}) must match "
                    f"out_dim ({self._out_dim})."
                )
            return self.eval(
                record, args=[] if args is None else args, active=active
            )

        def traverse(self, cb):
            cb.put(
                "network_weights",
                self.network_weights,
                mi.ParamFlags.Differentiable,
            )
            if self._encoding is not None:
                cb.put("encoding", self._encoding, mi.ParamFlags.Differentiable)

        def parameters_changed(self, keys=None):
            if dr.width(self.network_weights) != self._network_weight_count:
                raise RuntimeError(
                    "neuralfield: network_weights size does not match metadata."
                )

        def to_string(self):
            return (
                "NeuralField[\n"
                f"  out_dim = {self._out_dim},\n"
                f"  args_dim = {self._args_dim},\n"
                f'  decoder = "{self._decoder}"\n'
                "]"
            )

    return NeuralField
